File: solutions/javascript/003-luhn/solution.mjs
export function normalizeDigits(s) {
  return [...s].filter((ch) => ch !== " " && ch !== "-").join("");
}

export function luhnValid(raw) {
  const digits = normalizeDigits(raw);
  if (!digits || !/^\d+$/.test(digits)) return false;
  let sum = 0;
  let alt = false;
  for (let i = digits.length - 1; i >= 0; i--) {
    let d = digits.charCodeAt(i) - 48;
    if (alt) {
      d *= 2;
      if (d > 9) d -= 9;
    }
    sum += d;
    alt = !alt;
  }
  return sum % 10 === 0;
}

File: solutions/python/003-luhn/solution.py
"""Luhn checksum validation."""

def normalize_digits(s: str) -> str:
    return "".join(ch for ch in s if ch not in " -")


def luhn_valid(raw: str) -> bool:
    digits = normalize_digits(raw)
    if not digits or not digits.isdigit():
        return False
    total = 0
    alt = False
    for ch in reversed(digits):
        d = ord(ch) - 48
        if alt:
            d *= 2
            if d > 9:
                d -= 9
        total += d
        alt = not alt
    return total % 10 == 0
